### PR TITLE
Exponential confirm score

### PR DESCRIPTION
**Pull request by pieceofr**
Use the exponential formula for confirm-score 
- use  CDS data-set for Taiwan 
- use 14 days records  of confirm data to calculate the score
- use confirmDataCountThreadHold to control minimum data required

Formula link: https://github.com/bitmark-inc/autonomy-api/blob/master/share/jupyter/autonomyFormula.ipynb

### CODE REVIEW

File: store/metric.go [hunk 1]
```diff
@@ -100,7 +101,7 @@ func (m *mongoDB) CollectRawMetrics(location schema.Location) (*schema.Metric, e
 				TotalPeople:             totalPeopleReport,
 				MaxScorePerPerson:       schema.TotalOfficialBehaviorWeight,
 				CustomizedBehaviorTotal: float64(behaviorToday.CustomizedCount),
-				Score:                   behaviorScore,
+				Score: behaviorScore,
```
Comment (lemonlatte) [inline store/metric.go:104]: no change this

File: score/confirm.go [hunk 2]
```diff
@@ -1,24 +1,33 @@
...
 		return
 	}
-	// in between 0 - 10 people, each increased confirm case deduct 5 points
-	details.Score = 100 - 5*delta
-	if details.Score < 0 {
-		details.Score = 0
+	fraction := float64(0)
+	denominator := float64(0)
+	for idx, val := range dataset {
+		epx := (float64(idx) + 1) / 2
```
Comment (lemonlatte) [inline score/confirm.go:24]: `power` in the jupyter example is better than `epx`

File: score/confirm.go [hunk 3]
```diff
@@ -1,24 +1,33 @@
...
+	score := float64(0)
+	if len(dataset) < confirmDataCountThreadHold {
+		metric.Details.Confirm.Score = score
 		return
 	}
-	// in between 0 - 10 people, each increased confirm case deduct 5 points
-	details.Score = 100 - 5*delta
-	if details.Score < 0 {
-		details.Score = 0
+	fraction := float64(0)
```
Comment (nodestory) [inline score/confirm.go:21]:
```suggestion
	numerator := float64(0)
```

File: score/confirm.go [hunk 4]
```diff
@@ -1,24 +1,33 @@
...
-import "github.com/bitmark-inc/autonomy-api/schema"
+import (
+	"math"
+
+	"github.com/bitmark-inc/autonomy-api/schema"
+)
 
 const (
-	confirmCriteria = 10
+	confirmDataCountThreadHold = 5
```
Comment (nodestory) [inline score/confirm.go:10]:
```suggestion
	confirmDataCountThreadHold = 14
```

We use 14-day window for calculation so I suppose the threshold should be 14. If the number if data points is not sufficient, we consider it as 0.

File: store/confirmCDS.go [hunk 5]
```diff
@@ -83,57 +85,70 @@ func (m mongoDB) GetCDSConfirm(loc schema.Location) (float64, float64, float64,
...
-		today, delta, percent, err := m.GetConfirm(loc)
-		if err != nil {
-			return 0, 0, 0, fmt.Errorf("%s: %s", ErrConfirmDataFetch, err)
+		c = m.client.Database(m.database).Collection(CDSCountyCollectionMatrix[CDSCountryType(CdsTaiwan)])
+		opts := options.Find().SetSort(bson.M{"report_ts": -1}).SetLimit(2)
+		filter := bson.M{}
+		cur, err = c.Find(context.Background(), filter, opts)
+		if nil != err {
+			log.WithField("prefix", mongoLogPrefix).Errorf("CDS confirm data find  error: %s", err)
+			return 0, 0, 0, fmt.Errorf("cds confirm data find  error: %s", err)
```
Comment (nodestory) [inline store/confirmCDS.go:101]:
```suggestion
			return 0, 0, 0, ErrConfirmDataFetch
```

File: store/confirmCDS.go [hunk 6]
```diff
@@ -83,57 +85,70 @@ func (m mongoDB) GetCDSConfirm(loc schema.Location) (float64, float64, float64,
...
 		}
+		log.WithField("prefix", mongoLogPrefix).Debugf("cds query name: %s date:%s", result.Name, result.ReportTimeDate)
+		results = append(results, result)
 	}
-	return 0, 0, 0, ErrNoConfirmDataset
 
+	var percent, delta float64
+	var today, yesterday schema.CDSData
+	if len(results) >= 2 {
+		if results[0].ReportTime > results[1].ReportTime {
```
Comment (nodestory) [inline store/confirmCDS.go:128]: The query result is already sorted by `report_ts`, why is the time comparison needed here?

For the percent rate calculation, the util function is available.
https://github.com/bitmark-inc/autonomy-api/blob/master/score/utils.go#L3

File: store/confirmCDS.go [hunk 7]
```diff
@@ -164,3 +179,62 @@ func (m mongoDB) politicalGeoInfo(loc schema.Location) (PoliticalGeo, error) {
 	}
 	return ret, nil
 }
+func (m mongoDB) ContinuousDataCDSConfirm(loc schema.Location, num int64, lessEqThan int64) ([]schema.CDSScoreDataSet, int, error) {
```
Comment (nodestory) [inline store/confirmCDS.go:182]:
```suggestion
func (m mongoDB) ContinuousDataCDSConfirm(loc schema.Location, windowSize int64, latestReportTs int64) ([]schema.CDSScoreDataSet, error) {
```

File: store/confirmCDS.go [hunk 8]
```diff
@@ -164,3 +179,62 @@ func (m mongoDB) politicalGeoInfo(loc schema.Location) (PoliticalGeo, error) {
...
+		}
+		if len(now.Name) > 0 { // now data is valid
+			head := make([]schema.CDSScoreDataSet, 1)
+			head[0] = schema.CDSScoreDataSet{Name: now.Name, Cases: now.Cases - result.Cases, ReportTimeDate: now.ReportTimeDate}
+			results = append(head, results...)
+		}
+		now = result
+	}
+	cur.Close(ctx)
+	return results, len(results), nil
```
Comment (nodestory) [inline store/confirmCDS.go:239]:
```suggestion
	return results, nil
```

File: store/confirmCDS.go (hunk 6 above)
Comment (pieceofr) [reply store/confirmCDS.go:128]: They can be equal

File: store/confirmCDS.go (hunk 7 above)
Comment (pieceofr) [reply store/confirmCDS.go:182]: it is not alatestReportTs . it is more like a lessThan

File: store/confirmCDS.go (hunk 6 above)
Comment (pieceofr) [reply store/confirmCDS.go:128]: it can be equal

File: score/confirm.go (hunk 4 above)
Comment (pieceofr) [reply score/confirm.go]: remove threadhold

**Pull request merged**